src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,133 @@
+import classNames from 'classnames';
+import React, { useContext, useRef, useState } from 'react';
+
+import { Todo } from '../../types/Todo';
+import { TodosContext } from '../TodosContext';
+
+type Props = {
+  todo: Todo;
+};
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
+  const {
+    id, title, completed,
+  } = todo;
+
+  const {
+    isSubmiting,
+    deleteTodo,
+    updateTodo,
+  } = useContext(TodosContext);
+
+  const [isEdit, setIsEdit] = useState(false);
+  const [editTitle, setEditTitle] = useState(title);
+
+  const titleField = useRef<HTMLInputElement>(null);
+
+  const handleTodoDelete = (todoId : number) => {
+    deleteTodo(todoId);
+  };
+
+  const handleToggleTodo = () => {
+    updateTodo({ ...todo, completed: !completed });
+  };
+
+  const handleEditSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    const trimmedTitle = editTitle.trim();
+
+    switch (trimmedTitle) {
+      case '':
+        handleTodoDelete(id);
+        break;
+
+      case title:
+        setEditTitle(title);
+        break;
+
+      default:
+        updateTodo({ ...todo, title: trimmedTitle });
+        break;
+    }
+
+    setIsEdit(false);
+  };
+
+  const handleKeyUp = (event: React.KeyboardEvent) => {
+    if (event.key === 'Escape') {
+      setIsEdit(false);
+      setEditTitle(title);
+    }
+  };
+
+  const handleDoubleClick = () => {
+    setIsEdit(true);
+
+    setTimeout(() => {
+      titleField.current?.focus();
+    }, 1);
```

Comment:
why do you need this 1ms timeout?

Reply:
It's my mistake, timeout is not needed here


plus tempTodo was implemented in the function addTodo in the TodosContext component